### PR TITLE
Workflows input output

### LINKED ISSUE #556: Support TOSCA 1.3 inputs/outputs for workflows
**As:** an app designer
**I want to:** I want to provide some inputs when executing a workflow
**So that:** I can have different workflows executions without redeploying or updating an application

**AC1:** Should follow TOSCA 1.3, but if needed extensions are allowed
**AC2:** inputs should somehow be accessible to node operations (see spec)
**AC3:** If US #555 on workflows pre-conditions/conditions is already done allow to use inputs on it
**AC4:** Workflows outputs should be somehow retrievable from the REST API
**AC5:** Limited to Yorc, no support expected in Alien for now (will be a separated US).

### PR DESCRIPTION
**Pull request by laurentganne**
# Pull Request description

## Description of the change

Adding support for TOSCA 1.3 Workflow inputs and outputs.

### Workflow inputs

- Inputs can be defined in a workflow and will be provided in the execution context of the workflow.
- If an operation defines a get_input function on one of its parameter the input will be retrieved from the workflow input, and if not found from the topology inputs.

```yaml
  workflows:
    compute_pi:
      inputs:
        precision:
          type: int
          required: true
```

Yorc CLI `yorc deployments workflows execute` and Yorc REST API `/deployments/<deployment_id>/workflows/<workflow_name>` were updated to allow to provide input parameters when executing a workflow.

### Workflow outputs
- Allow to specify **workflow outputs** as **Tosca Parameters**  (cf topology outputs)
- The workflow outputs are saved in the corresponding **Task** bound to the workflow execution. 

```yaml
  workflows:
    compute_pi:
      outputs:
        pi:
          description: The PI number calculation result
          value: { get_attribute: [ ComputePIComponentType, result ] }
```

### Operation outputs attribute mappings
- Allow to specify operation outputs as **Tosca Attribute mapping**

```yaml
    interfaces:
      custom:
        compute_pi:
          outputs:
            result: [SELF, result]
          implementation: scripts/compute_pi.py
```

### How to verify it
As not yet implemented in Alien4Cloud, the test only involves yorc.

#### Deploy the following test application [deployment.zip](https://github.com/ystia/yorc/files/4307971/deployment.zip)

```bash
$ yorc deployments deploy deployment.zip  --id test
````
Check logs or run this command to wait until the end of deployment:

```bash
$ yorc deployments info -f test
```

#### Check the compute_pi workflow, what are its expected inputs and outputs:

Run this command to verify the workflow expects in input a precision,
and provides in ouput the value of pi:

```bash
$ yorc deployments workflows show test -w compute_pi
Workflow compute_pi:
  Inputs:
    precision:
      type: int
      required: true
  Step PiComponent_compute_pi:
    Target: PiComponent
    Activities:
      - Call Operation: custom.compute_pi
  Outputs:
    pi:
      description: PI computation result
```

#### Run compute_pi with a precision of 3 digits

Run this command, specifying the workflow input parameter **precision** has the value **3**:

```bash 
$ yorc deployments workflows execute test -w compute_pi \
   --data '{"inputs":{"precision":3}}'

New task  9ffe4c0b-8c74-4697-8eb9-c5fffe551707  created to execute  compute_pi
```

This command has returned a task ID **9ffe4c0b-8c74-4697-8eb9-c5fffe551707**,
that will be used below to retrieve the workflow output.

Check yorc log to see logs of the task execution.
Once done, run the following command to get info on the task

```bash 
$ yorc d tasks info test 9ffe4c0b-8c74-4697-8eb9-c5fffe551707
Task:  9ffe4c0b-8c74-4697-8eb9-c5fffe551707
Task status: DONE
Task type: CustomWorkflow
Task outputs:
+------+-------+
| Name | Value |
+------+-------+
| pi   | 3.142 |
+------+-------+
```

The workflow output value has the expected precision.

#### Run again compute_pi, with a precision of 5 digits this time

Run this command, specifying the workflow input parameter **precision** has the value **5**:

```bash
$ yorc deployments workflows execute test -w compute_pi \
   --data '{"inputs":{"precision":5}}'

New task  ce7c8697-fecf-40f1-add1-2d98a81bd416  created to execute  compute_pi
```

Once the task is done, run this command to check the workflow output:

```bash
$ yorc d tasks info test ce7c8697-fecf-40f1-add1-2d98a81bd416
Task:  ce7c8697-fecf-40f1-add1-2d98a81bd416
Task status: DONE
Task type: CustomWorkflow
Task outputs:
+------+---------+
| Name | Value   |
+------+---------+
| pi   | 3.14159 |
+------+---------+
```

The output has a precision of 5 digits.

### Description for the changelog

Support TOSCA 1.3 inputs/outputs for workflows ([GH-556](https://github.com/ystia/yorc/issues/556))

## Applicable Issues

Closes #556

### CODE REVIEW

File: deployments/datatypes.go [hunk 1]
```diff
@@ -162,3 +162,12 @@ func getDataTypeComplexType(dataType string) string {
 	}
 	return tType
 }
+
+func isPrimitiveDataType(dataType string) bool {
```
Comment (loicalbertin) [inline deployments/datatypes.go]: Why not use https://github.com/ystia/yorc/blob/develop/tosca/types.go#L36  here?

File: deployments/workflows.go [hunk 2]
```diff
@@ -116,3 +119,32 @@ func enhanceWorkflows(ctx context.Context, deploymentID string) error {
...
+	wf, err := GetWorkflow(ctx, deploymentID, workflowName)
+	if err != nil {
+		return nil, err
+	}
+
+	outputs := make(map[string]*TOSCAValue)
+	for outputName, outputDef := range wf.Outputs {
+		dataType := getTypeFromParamDefinition(ctx, &outputDef)
+
+		res, err := getValueAssignment(ctx, deploymentID, "", "0", "", dataType, outputDef.Value)
```
Comment (loicalbertin) [inline deployments/workflows.go]: So here we always resolve on instance "0"?

File: prov/ansible/execution.go [hunk 3]
```diff
@@ -20,6 +20,7 @@ import (
 	"context"
 	"encoding/csv"
 	"fmt"
+	"github.com/ystia/yorc/v4/helper/consulutil"
```
Comment (loicalbertin) [inline prov/ansible/execution.go]: misplaced

File: prov/ansible/execution.go [hunk 4]
```diff
@@ -573,52 +573,95 @@ func (e *executionCommon) resolveOperationOutputPath() error {
...
-		oof := va.GetFunction()
-		if oof.Operator != tosca.GetOperationOutputOperator {
-			return errors.Errorf("Output %q for operation %v (%v) is not a valid get_operation_output TOSCA function", outputName, e.operation, oof)
-		}
-		targetContext := oof.Operands[0].String() == "TARGET"
-		sourceContext := oof.Operands[0].String() == "SOURCE"
-		if (targetContext || sourceContext) && !e.operation.RelOp.IsRelationshipOperation {
-			return errors.Errorf("Can't resolve an operation output in SOURCE or TARGET context without a relationship operation: %q", va.String())
+		// either output can be associated to a get_operation_output value assignment or to an attribute mapping
+		if outputValue.ValueAssign != nil {
```
Comment (loicalbertin) [inline prov/ansible/execution.go]: Not in this PR but we should definitively think about extracting this from the ansible package!

File: prov/ansible/execution.go [hunk 5]
```diff
@@ -1084,31 +1127,44 @@ func (e *executionCommon) executeWithCurrentInstance(ctx context.Context, retry
...
+						attributeOrCapabilityName := data[3]
+						var parameters []string
+						if len(data) > 3 {
+							parameters = data[4:]
+						}
+						if err = deployments.ResolveAttributeMapping(ctx, e.deploymentID, nodeName, instanceName, attributeOrCapabilityName, line[1], parameters...); err != nil {
+							return err
+						}
+					} else {
+						// TODO this should be part of the deployments package
```
Comment (loicalbertin) [inline prov/ansible/execution.go]: Still relevant :smile:

File: rest/http_api.md [hunk 6]
```diff
@@ -680,6 +680,11 @@ By default the execution of the workflow's steps take place on all the instances
 It is possible to select instances for the workflow's nodes by adding selection data in the request body.
 For nodes that have no selected instances specified, the execution steps take place on all instances.
 
+The request bdoy can also contain input values assignments, that will be provided
```
Comment (loicalbertin) [inline rest/http_api.md]:
```suggestion
The request body can also contain input values assignments, that will be provided
```

Comment (laurentganne) [reply rest/http_api.md]: good catch, will fix

**Comment by laurentganne** [comment]
> 
> 
> Is there a reason why we do not check if there is required inputs when requesting for running a custom workflow?
> I mean before submitting a task that will fail.

Proposing to implement this check in the REST API:
if the request does not contain an workflow input parameter described as required in the TOSCA description of the workflow, then return a failure instead of creating the task and let the user discover the issue in the task execution logs

**Comment by laurentganne** [comment]
Added a check in REST code to fail if the request body does not provide a value for a required workflow input parameter.

File: deployments/attribute_mappings.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+	"context"
+	"github.com/pkg/errors"
+	"github.com/ystia/yorc/v4/helper/collections"
+	"strconv"
+)
+
+// ResolveAttributeMapping allows to resolve an attribute mapping
+// i.e update the corresponding attribute with the defined value
+// parameters can be nested keys and/or attribute name in case of capability
+func ResolveAttributeMapping(ctx context.Context, deploymentID, nodeName, instanceName, capabilityOrAttributeName string, attributeValue interface{}, parameters ...string) error {
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply deployments/attribute_mappings.go]: Tosca spec

File: deployments/datatypes.go (hunk 1 above)
Comment (stebenoist) [reply deployments/datatypes.go]: Didn't see this but seems weird map and list are considered like builtin type if entry shcema is not a primitive

File: deployments/workflows.go (hunk 2 above)
Comment (stebenoist) [reply deployments/workflows.go]: Yes, workflow output resolution is not dependent to instances. Am I wrong ?

File: prov/ansible/execution.go (hunk 3 above)
Comment (stebenoist) [reply prov/ansible/execution.go]: Fixed

File: prov/ansible/execution.go (hunk 4 above)
Comment (stebenoist) [reply prov/ansible/execution.go]: Yes !

File: prov/ansible/execution.go (hunk 5 above)
Comment (stebenoist) [reply prov/ansible/execution.go]: Yes !

File: deployments/attribute_mappings.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (stebenoist) [reply deployments/attribute_mappings.go]: I added some comments to clarify it

File: deployments/datatypes.go (hunk 1 above)
Comment (stebenoist) [reply deployments/datatypes.go]: Fixed

File: deployments/attribute_mappings.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (loicalbertin) [reply deployments/attribute_mappings.go]: looks good

**Pull request merged**